pints/tests/test_residuals_diagnostics.py

```diff
@@ -0,0 +1,199 @@
+import matplotlib
+
+# Select matplotlib backend that doesn't require a screen
+matplotlib.use('Agg')
+
+
+class TestResidualsDiagnostics(unittest.TestCase):
+    """
+    Tests Pints residuals diagnostics methods.
+    """
+    def __init__(self, name):
+        super(TestResidualsDiagnostics, self).__init__(name)
+
+        # Create a single output optimization toy model
+        self.model1 = toy.LogisticModel()
+        self.real_parameters1 = [0.015, 500]
+        self.times1 = np.linspace(0, 1000, 100)
+        self.values1 = self.model1.simulate(self.real_parameters1, self.times1)
+
+        # Add noise
+        self.noise1 = 50
+        self.values1 += np.random.normal(0, self.noise1, self.values1.shape)
+
+        # Set up optimisation problem
+        self.problem1 = pints.SingleOutputProblem(
+            self.model1, self.times1, self.values1)
+        self.score1 = pints.SumOfSquaresError(self.problem1)
+        self.boundaries1 = pints.RectangularBoundaries([0, 200], [1, 1000])
+        self.x01 = np.array([0.5, 500])
+
+        # Run the optimisation
+        optimiser = pints.OptimisationController(
+            self.score1,
+            self.x01,
+            boundaries=self.boundaries1,
+            method=pints.XNES,
+        )
+        optimiser.set_log_to_screen(False)
+        self.found_parameters1, self.found_value1 = optimiser.run()
+
+        # Create a multiple output MCMC toy model
+        self.model2 = toy.LotkaVolterraModel()
+        self.real_parameters2 = self.model2.suggested_parameters()
+        # Downsample the times for speed
+        self.times2 = self.model2.suggested_times()[::10]
+        self.values2 = self.model2.simulate(self.real_parameters2, self.times2)
+
+        # Add noise
+        self.noise2 = 0.05
+        self.values2 += np.random.normal(0, self.noise2, self.values2.shape)
+
+        # Create an object with links to the model and time series
+        self.problem2 = pints.MultiOutputProblem(
+            self.model2, self.times2, self.values2)
+
+        # Create a uniform prior over both the parameters and the new noise
+        # variable
+        self.log_prior2 = pints.UniformLogPrior([1, 1, 1, 1], [6, 6, 6, 6])
+        # Create a log likelihood
+        self.log_likelihood2 = pints.GaussianKnownSigmaLogLikelihood(
+            self.problem2, self.noise2)
+
+        # Create an un-normalised log-posterior (log-likelihood + log-prior)
+        self.log_posterior2 = pints.LogPosterior(
+            self.log_likelihood2, self.log_prior2)
+
+        # Run MCMC
+        self.x02 = [
+            self.real_parameters2 * 1.1,
+            self.real_parameters2 * 0.9,
+            self.real_parameters2 * 1.05
+        ]
+        mcmc = pints.MCMCController(self.log_posterior2, 3, self.x02)
+        mcmc.set_max_iterations(300)  # make it as small as possible
+        mcmc.set_log_to_screen(False)
+        self.samples2 = mcmc.run()
+
+    def test_calculate_residuals(self):
+        """
```

Comment:
There's an annoying thing with unit testing where it doesn't show filename or method name if your test has a docstring, so we try to _not_ have docstrings in tests, but instead just add comments at the top (e.g. `# Test the calculate_residuals function.`)